### PR TITLE
rgw: fix lockdep false positive

### PR DESCRIPTION
**Pull request by yehudasa**
Signed-off-by: Yehuda Sadeh <yehuda@redhat.com>

### CODE REVIEW

File: src/rgw/rgw_data_sync.cc [hunk 1]
```diff
@@ -1239,17 +1242,16 @@ class RGWDataSyncCR : public RGWCoroutine {
...
-                 iter != sync_status.sync_markers.end(); ++iter) {
-              RGWDataSyncShardControlCR *cr = new RGWDataSyncShardControlCR(sync_env, sync_env->store->get_zone_params().log_pool,
-                                                        iter->first, iter->second);
-              cr->get();
-              shard_crs_lock.Lock();
-              shard_crs[iter->first] = cr;
-              shard_crs_lock.Unlock();
-              spawn(cr, true);
-            }
+          for (map<uint32_t, rgw_data_sync_marker>::iterator iter = sync_status.sync_markers.begin();
```
Comment (theanalyst) [inline src/rgw/rgw_data_sync.cc]: nit: cant we use auto here?

File: src/rgw/rgw_data_sync.cc [hunk 2]
```diff
@@ -1239,17 +1242,16 @@ class RGWDataSyncCR : public RGWCoroutine {
 
       yield {
         if  ((rgw_data_sync_info::SyncState)sync_status.sync_info.state == rgw_data_sync_info::StateSync) {
-          case rgw_data_sync_info::StateSync:
```
Comment (oritwas) [inline src/rgw/rgw_data_sync.cc]: why was this removed?

File: src/rgw/rgw_data_sync.cc [hunk 3]
```diff
@@ -1182,9 +1182,12 @@ class RGWDataSyncCR : public RGWCoroutine {
   }
 
   ~RGWDataSyncCR() {
+    shard_crs_lock.Lock();
```
Comment (cbodley) [inline src/rgw/rgw_data_sync.cc]: shouldn't need to lock in a dtor - if we get here while operate() is still running, that would indicate a separate ref counting issue. did you see this happening?

File: src/rgw/rgw_data_sync.cc (hunk 1 above)
Comment (yehudasa) [inline src/rgw/rgw_data_sync.cc]: @theanalyst yes we can, but this is just indentation fix (and removed extra line), so I wouldn't go into changing it just now

File: src/rgw/rgw_data_sync.cc (hunk 2 above)
Comment (yehudasa) [inline src/rgw/rgw_data_sync.cc]: @oritwas this line shouldn't have been there.. there's no switch statement, it only compiles because the reenter/yield expand into a switch statement. The `if` there replaced the original switch.

File: src/rgw/rgw_data_sync.cc (hunk 3 above)
Comment (yehudasa) [inline src/rgw/rgw_data_sync.cc]: @cbodley you have a point, I've seen shard_crs inconsistency, but can't be sure what the culprit was.

File: src/rgw/rgw_data_sync.cc (hunk 2 above)
Comment (cbodley) [inline src/rgw/rgw_data_sync.cc]: it looks like this used to be a switch statement that changed to an if on the line above. i assume the compiler didn't complain about the leftover `case` here because the boost `reenter` macro is using a switch under the covers

File: src/rgw/rgw_data_sync.cc (hunk 1 above)
Comment (theanalyst) [inline src/rgw/rgw_data_sync.cc]: Ah ok, no issues

File: src/rgw/rgw_data_sync.cc (hunk 3 above)
Comment (yehudasa) [inline src/rgw/rgw_data_sync.cc]: @cbodley removed this

**Comment by yehudasa** [comment]
@cbodley added a couple more fixes, and made the change according to your comment
